File: src/services/toolDetectionService.ts
import { exec } from 'child_process';
import { promisify } from 'util';

const execAsync = promisify(exec);

export interface ToolStatus {
  name: string;
  executable: string;
  isAvailable: boolean;
  version: string;
  installGuideUrl: string;
}

export class ToolDetectionService {
  private readonly tools = [
    { name: 'MySQL Dump', executable: 'mysqldump', versionFlag: '--version', guide: 'https://dev.mysql.com/downloads/installer/' },
    { name: 'PostgreSQL Dump', executable: 'pg_dump', versionFlag: '--version', guide: 'https://www.postgresql.org/download/' },
    { name: 'SQL Server Command', executable: 'sqlcmd', versionFlag: '-?', guide: 'https://learn.microsoft.com/en-us/sql/tools/sqlcmd-utility' },
    { name: 'SQLite3 CLI', executable: 'sqlite3', versionFlag: '--version', guide: 'https://www.sqlite.org/download.html' },
    { name: 'Oracle Export', executable: 'exp', versionFlag: 'help=y', guide: 'https://www.oracle.com/database/technologies/instant-client/downloads.html' }
  ];

  public async getToolsStatus(): Promise<ToolStatus[]> {
    const results: ToolStatus[] = [];

    for (const tool of this.tools) {
      try {
        const { stdout, stderr } = await execAsync(`"${tool.executable}" ${tool.versionFlag}`);
        const output = (stdout || stderr).trim();
        const firstLine = output.split('\n')[0];

        results.push({
          name: tool.name,
          executable: tool.executable,
          isAvailable: true,
          version: firstLine.length > 50 ? firstLine.substring(0, 50) + '...' : firstLine,
          installGuideUrl: tool.guide
        });
      } catch {
        results.push({
          name: tool.name,
          executable: tool.executable,
          isAvailable: false,
          version: '未偵測到此 CLI 工具',
          installGuideUrl: tool.guide
        });
      }
    }

    return results;
  }
}

File: backend/MyNavicat.Api/Services/ToolDetectionService.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading.Tasks;
using MyNavicat.Api.Models.Common;

namespace MyNavicat.Api.Services
{
    public interface IToolDetectionService
    {
        Task<List<ToolStatus>> DetectToolsAsync();
        bool IsToolAvailable(string toolName);
    }

    /// <summary>
    /// 資料庫 CLI 工具偵測服務實作
    /// </summary>
    public class ToolDetectionService : IToolDetectionService
    {
        private static readonly List<(string Name, string DisplayName, string TargetDb, string VersionArg, string InstallUrl)> RegisteredTools = new()
        {
            ("mysqldump", "MySQL Dump Tool", "MySQL", "--version", "https://dev.mysql.com/doc/refman/8.0/en/mysqldump.html"),
            ("pg_dump", "PostgreSQL Dump Tool", "PostgreSQL", "--version", "https://www.postgresql.org/docs/current/app-pgdump.html"),
            ("sqlcmd", "SQL Server Command Line Utility", "SqlServer", "-?", "https://learn.microsoft.com/en-us/sql/tools/sqlcmd/sqlcmd-utility"),
            ("sqlite3", "SQLite3 Command Line Interface", "SQLite", "-version", "https://www.sqlite.org/cli.html"),
            ("exp", "Oracle Export Utility", "Oracle", "-help", "https://docs.oracle.com/en/database/oracle/oracle-database/19/xeaiw/index.html")
        };

        public async Task<List<ToolStatus>> DetectToolsAsync()
        {
            var results = new List<ToolStatus>();

            foreach (var tool in RegisteredTools)
            {
                var status = await CheckToolAsync(tool.Name, tool.DisplayName, tool.TargetDb, tool.VersionArg, tool.InstallUrl);
                results.Add(status);
            }

            return results;
        }

        public bool IsToolAvailable(string toolName)
        {
            try
            {
                using var process = new Process
                {
                    StartInfo = new ProcessStartInfo
                    {
                        FileName = toolName,
                        Arguments = "--version",
                        RedirectStandardOutput = true,
                        RedirectStandardError = true,
                        UseShellExecute = false,
                        CreateNoWindow = true
                    }
                };

                return process.Start();
            }
            catch
            {
                return false;
            }
        }

        private async Task<ToolStatus> CheckToolAsync(string name, string displayName, string targetDb, string versionArg, string installUrl)
        {
            var status = new ToolStatus
            {
                Name = name,
                DisplayName = displayName,
                TargetDbType = targetDb,
                InstallGuideUrl = installUrl,
                IsAvailable = false,
                Version = null
            };

            try
            {
                using var process = new Process
                {
                    StartInfo = new ProcessStartInfo
                    {
                        FileName = name,
                        Arguments = versionArg,
                        RedirectStandardOutput = true,
                        RedirectStandardError = true,
                        UseShellExecute = false,
                        CreateNoWindow = true
                    }
                };

                if (process.Start())
                {
                    string output = await process.StandardOutput.ReadToEndAsync();
                    string error = await process.StandardError.ReadToEndAsync();
                    await process.WaitForExitAsync();

                    string fullText = string.IsNullOrWhiteSpace(output) ? error : output;

                    if (!string.IsNullOrWhiteSpace(fullText))
                    {
                        var firstLine = fullText.Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries)[0];
                        status.Version = firstLine.Length > 100 ? firstLine.Substring(0, 100) : firstLine;
                        status.IsAvailable = true;
                    }
                    else
                    {
                        status.IsAvailable = true;
                        status.Version = "Installed (Version unknown)";
                    }
                }
            }
            catch
            {
                status.IsAvailable = false;
                status.Version = null;
            }

            return status;
        }
    }
}
